### PR TITLE
Tests for concatMap operator

### PR DESCRIPTION
**Pull request by arkivanov**
Also improved `flatMap` tests a bit (and removed one useless test)

### CODE REVIEW

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        val inner = TestObservable<String>()
+        testConcatMap { inner }
+        upstream.onNext(0)
+
+        inner.onError(Throwable())
+
+        assertTrue(upstream.isDisposed)
+    }
+
+    private fun testConcatMap(innerSources: List<Observable<String?>>): TestObservableObserver<String?> =
```
Comment (mk-jones) [inline reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt]: It's not clear what these methods do. Perhaps subscribeWithIndexedObservableList?

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        testConcatMap { innerSources[it!!] }
+
+    private fun testConcatMap(mapper: (Int?) -> Observable<String?>): TestObservableObserver<String?> =
+        upstream.concatMap(mapper).test()
+
+    private fun innerSources(count: Int): List<TestObservable<String?>> =
```
Comment (mk-jones) [inline reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt]: createTestObservables?

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+        inner.onError(Throwable())
+
+        assertTrue(upstream.isDisposed)
+    }
+
+    private fun testConcatMap(innerSources: List<Observable<String?>>): TestObservableObserver<String?> =
+        testConcatMap { innerSources[it!!] }
+
+    private fun testConcatMap(mapper: (Int?) -> Observable<String?>): TestObservableObserver<String?> =
```
Comment (mk-jones) [inline reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt]: subscribeWithMapper?

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+        assertTrue(inners[1].hasSubscribers)
+    }
+
+    @Test
+    fun produces_values_in_correct_order() {
+        val inners = innerSources(3)
+        val observer = testConcatMap() { inners[it ?: 0] }
+
+        upstream.onNext(null)
```
Comment (mk-jones) [inline reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/ConcatMapTest.kt]: Would be nice if we had some way of making this test more legible. However I'm out of ideas.

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/FlatMapTest.kt [hunk 5]
```diff
@@ -188,6 +179,12 @@ class FlatMapTest : UpstreamDownstreamGenericTests by UpstreamDownstreamGenericT
         assertTrue(inners[1].isDisposed)
     }
 
-    private fun testFlatMap(mapper: (Int) -> Observable<String>): TestObservableObserver<String> =
+    private fun testFlatMap(innerSources: List<Observable<String?>>): TestObservableObserver<String?> =
```
Comment (mk-jones) [inline reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/FlatMapTest.kt]: Ditto to concatMap.

**Comment by arkivanov** [comment]
@mk-jones Fixed, please check

**Comment by arkivanov** [comment]
@mk-jones Resolved conflicts, needs another approval

**Pull request merged**